File: python/minimizer.py
import sys, time, resource

# read console arguments
input_file = sys.argv[1]
w = int(sys.argv[2])
k = int(sys.argv[3])
l = w+k-1

# array to store minimizers
minimizers = []
    
# function 'find minimizer within a substring'
def minimizer(str):
  array = []

  for i in range (0, len(str)-k+1):
    array.append(str[i:i+k])
    
  return min(array)

# function 'find left-end minimizers'
def left_end(str):
  array = []
  
  for i in range (k, l):
    min_seed = minimizer(str[0:i])
    array.append(min_seed)
    
  return array
  
# function 'find right-end minimizers'
def right_end(str):
  array = []
  
  for i in range (0, l-k):
    min_seed = minimizer(str[i:l])
    array.append(min_seed)
    
  return array
  
# function 'remove duplicates'
def remove_duplicates(array):
  new_array = []
  
  for str in array:
    if (not new_array or new_array[-1] != str):
      new_array.append(str)
      
  return new_array
  
# ---------------------------------main------------------------------------------

time_start = time.clock()

# read input file
with open (input_file, "r") as f_in:
  sequence = f_in.read().replace('\r', '').replace('\n', '')

seq_len = len(sequence)
  
# find left-end minimizers
left_minimizers = left_end(sequence[0:l])
for str in left_minimizers:
  minimizers.append(str)

# find interior minimizers
for i in range (0, seq_len-l+1):
  min_seed = minimizer(sequence[i:i+l])
  minimizers.append(min_seed)
  
# find right-end minimizers
right_minimizers = right_end(sequence[seq_len-l+1:seq_len])
for str in right_minimizers:
  minimizers.append(str)
  
# remove duplicates
minimizers = remove_duplicates(minimizers)

# write minimizers to file
f_out = open('output.txt','w')
for item in minimizers:
  f_out.write("%s\n" % item)
f_out.close()

# calculate time and memory consumption
duration = time.clock() - time_start
memory  = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
  
# print out variables
print "Finished!"
print "Duration: ", duration, "seconds"
print "Memory consumption: ",  memory,  "KB"

File: cpp/minimizer.cpp
#include <vector>
#include <string>
#include <iostream>
#include <fstream>
#include <algorithm> //for str.erase
#include <set>
#include <iterator>

int w;
int k;
int l;

//minimizer() function
std::string minimizer(std::string subString){
  std::set<std::string> sortedItems;
  for(int i=0; i<(int)subString.size()-k+1; i++){
    sortedItems.insert(subString.substr(i,k));
    //std::cout << subString.substr(i,k) << std::endl;
  }
  //std::cout << *sortedItems.begin() << std::endl;

  return *sortedItems.begin();
}

//left_end() function
std::vector<std::string> left_end(std::string subString){
  std::vector<std::string> leftEndSeeds;
  for(int i=k; i<l; i++){
    leftEndSeeds.push_back(minimizer(subString.substr(0,i)));
    //std::cout << subString.substr(0,i) << std::endl;
  }
  return leftEndSeeds;
}

//right_end() function
std::vector<std::string> right_end(std::string subString){
  std::vector<std::string> rightEndSeeds;
  for(int i=0; i<l-k; i++){
    rightEndSeeds.push_back(minimizer(subString.substr(i,l-1)));
    //std::cout << subString.substr(i,l-1) << std::endl;
  }
  //for(auto i=rightEndSeeds.begin(); i!=rightEndSeeds.end(); ++i){
  //  std::cout << *i << std::endl;
  //}
  return rightEndSeeds;
}

//main() function
int main(int argc, char *argv[])
{
  //Init
  char* inputFile = argv[1];
  std::vector<std::string> minimizers;
  std::vector<std::string> leftEndMin;
  std::vector<std::string> rightEndMin;
  w = (int)(argv[2][0] - '0');
  k = (int)(argv[3][0] - '0');
  l = w+k-1;

  std::cout << "w=" << w << ", k=" << k << ", l=" << l << std::endl;

  //Loading input file to string and removing \n and \r
  std::ifstream ifs(inputFile);
  std::string genome( (std::istreambuf_iterator<char>(ifs) ),
                       (std::istreambuf_iterator<char>()    ) );  
  genome.erase(std::remove(genome.begin(), genome.end(), '\n'), genome.end());
  genome.erase(std::remove(genome.begin(), genome.end(), '\r'), genome.end());
  
  std::cout << "Loaded " << genome.size() << " chars from string!" << std::endl;
  //std::cout << genome << std::endl;

  //Find left end minimizers
  leftEndMin = left_end(genome.substr(0,l-1));
  //std::cout << "leftEndMin: " << leftEndMin.size() << std::endl;
  for(int i=0; i<(int)leftEndMin.size(); i++){
    minimizers.push_back(leftEndMin[i]);
  }

  //Find interior minimizers
  for(int i=0; i<(int)genome.size()-l+1; i++){
    //std::cout << minimizer(genom.substr(i,k+w-1)) << std::endl;
    minimizers.push_back(minimizer(genome.substr(i,l)));
  }

  //Find right end minimizers
  rightEndMin = right_end(genome.substr(genome.size()-l+1,l-1));
  //std::cout << "rightEndMin: " << rightEndMin.size() << std::endl;
  for(int i=0; i<(int)rightEndMin.size(); i++){
    minimizers.push_back(rightEndMin[i]);
  }

  //Delete duplicates and print to stdout
  minimizers.erase(unique(minimizers.begin(), minimizers.end()), minimizers.end());
  //for(auto i=minimizers.begin(); i!=minimizers.end(); ++i){
  //  std::cout << *i << std::endl;
  //}

  //Write the results to a text file
  std::ofstream output_file("./output.txt");
  std::ostream_iterator<std::string> output_iterator(output_file, "\n");
  std::copy(minimizers.begin(), minimizers.end(), output_iterator);
  std::cout << "Stored " << minimizers.size() << " minimizers to output.txt" << std::endl;
  
  return 0;
}
